pkg/core/klinebuilder.go

```diff
@@ -0,0 +1,207 @@
+		}
+	} else {
+		c.high = c.price
+	}
+	if !c.low.IsZero() {
+		if c.price.Compare(c.low) < 0 {
+			c.low = c.price
+		}
+	} else {
+		c.low = c.price
+	}
+	if c.open.IsZero() {
+		c.open = c.price
+	}
+	c.volume = c.volume.Add(trade.Quantity)
+	c.quoteVolume = c.quoteVolume.Add(trade.QuoteQuantity)
+	c.numTrades++
+	c.updateTime = trade.Time
+}
+
+func (c *kLineBuilder) Run(ctx context.Context) {
+	if len(c.intervals) == 0 || c.isBackTesting {
+		return
+	}
+	minDuration := c.minInterval.Duration()
+	for interval := range c.intervals {
+		if interval.Duration()%minDuration != 0 {
+			err := fmt.Errorf("interval %s is not a multiple of minInterval %s", interval, c.minInterval)
+			c.logger.Error(err)
+			panic(err)
+		}
+	}
+	c.logger.Infof("kline updater started for %s (%+v)", c.symbol, c.intervals)
+	go c.update(ctx)
+}
+
+func (c *kLineBuilder) update(ctx context.Context) {
+	// wait for the next interval
+	startTime := time.Now()
+	c.updateTime = types.Time(startTime) // kick-start update time
+	startTimeRound := startTime.Round(c.minInterval.Duration())
+	waitDuration := startTimeRound.Sub(startTime)
+	if waitDuration > 0 {
+		select {
+		case <-time.After(waitDuration):
+		case <-ctx.Done():
+			return
+		}
+	}
+	// start ticker
+	intervalDuration := c.minInterval.Duration()
+	ticker := time.NewTicker(intervalDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case tickTime := <-ticker.C:
+			startTime := tickTime.Add(-1 * intervalDuration).Round(intervalDuration)
+
+			// construct kline and reset converter
+			c.mu.Lock()
+			lastTradeTime := c.updateTime
+			klineMinInterval := types.KLine{
+				Symbol:    c.symbol,
+				StartTime: types.Time(startTime),
+				EndTime:   types.Time(tickTime),
+				Interval:  c.minInterval,
+				Closed:    true,
+			}
+			klineMinInterval.Open = c.open
+			klineMinInterval.Close = c.close
+			klineMinInterval.High = c.high
+			klineMinInterval.Low = c.low
+			klineMinInterval.Volume = c.volume
+			klineMinInterval.QuoteVolume = c.quoteVolume
+			klineMinInterval.NumberOfTrades = c.numTrades
+			c.open = fixedpoint.Zero
+			c.close = fixedpoint.Zero
+			c.high = fixedpoint.Zero
+			c.low = fixedpoint.Zero
+			c.volume = fixedpoint.Zero
+			c.quoteVolume = fixedpoint.Zero
+			c.numTrades = 0
+			c.mu.Unlock()
+
+			// update klines map
+			c.updateKLinesMap(&klineMinInterval, lastTradeTime)
+		}
+	}
+}
+
+func (c *kLineBuilder) updateKLinesMap(klineMinInterval *types.KLine, lastUpdateTime types.Time) {
+	for interval := range c.intervals {
+		if interval == c.minInterval {
+			c.EmitKLineClosed(*klineMinInterval)
+			continue
+		}
+		kline, ok := c.klinesMap[interval]
+		if !ok {
+			// interval not in map and kline is nil
+			kline = &types.KLine{}
+			kline.Set(klineMinInterval)
+			c.klinesMap[interval] = kline
+		} else {
+			kline.Merge(klineMinInterval)
+		}
+		kline.Interval = interval
+		expectEndTime := kline.StartTime.Time().Add(interval.Duration())
```

Comment:
-1 time.millisecond?

Reply:
I believe we don't need to do that here.

Ideally, `klineMinInterval` should end at a timestamp which is a multiple of `minInterval`.
`expectEndTime` here is the ideal end time of the `klineMinInterval`.

So basically, L181 is checking:
1. if the `klineMinInterval.EndTime`, which in fact is the ticker time, is after `expectEndTime`.
2. or the `lastUpdateTime`, which is the time of the last market trade, is after `expectEndTime`.

Either case, the kline should be closed for that interval.